### PR TITLE
Cover more cases for node module auto-import

### PR DESCRIPTION
**Pull request by hanzooo**
Fixes https://github.com/microsoft/TypeScript/issues/54023

handle the case that use a directory  as the package.json "main" field

### CODE REVIEW

**Comment by hanzooo** [comment]
> This PR doesn't have any linked issues. Please open an issue that references this PR. From there we can discuss and prioritise.

@typescript-bot check it again？

**Review by andrewbranch — CHANGES_REQUESTED** [review]
This is close, but needs a couple more conditions. This behavior is deprecated for packages with `"type": "module"`:

image

And TypeScript module resolution doesn’t support it. So we can’t do this if the package has `"type": "module"`.

Secondly, we need to make sure that the extension on `index.{???}` is something that supports omitting the extension. E.g., if the file is `index.mjs` or `index.d.cts`, those cannot be resolved by a directory reference.

**Comment by hanzooo** [comment]
> 

@andrewbranch Thx for the review, and detailed feedback~ I just added more checks for this.

File: src/compiler/moduleSpecifiers.ts [hunk 1]
```diff
@@ -1007,10 +1008,24 @@ function tryGetModuleNameAsNodeModule({ path, isRedirect }: ModulePath, { getCan
...
                 const mainExportFile = toPath(mainFileRelative, packageRootPath, getCanonicalFileName);
-                if (removeFileExtension(mainExportFile) === removeFileExtension(getCanonicalFileName(moduleFileToTry))) {
+                const canonicalModuleFileToTry = getCanonicalFileName(moduleFileToTry);
+                if (removeFileExtension(mainExportFile) === removeFileExtension(canonicalModuleFileToTry)) {
                     // ^ An arbitrary removal of file extension for this comparison is almost certainly wrong
                     return { packageRootPath, moduleFileToTry };
                 }
+                else if (
+                    packageJsonContent.type !== "module" &&
+                    !fileExtensionIsOneOf(canonicalModuleFileToTry, [Extension.Mjs, Extension.Cjs, Extension.Dmts, Extension.Dcts]) &&
```
Comment (andrewbranch) [inline src/compiler/moduleSpecifiers.ts:1018]:
```suggestion
                    !fileExtensionIsOneOf(canonicalModuleFileToTry, extensionsNotSupportingExtensionlessResolution) &&
```

(needs import too)

**Review by andrewbranch — APPROVED** [review]
Thank you!

**Pull request merged**